File: Data Structures & Algorithms/count-number-of-islands/submission-4.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        islands = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    islands += 1
                    self.sinkIsland(grid, r, c, rows, cols)
        
        return islands

    def sinkIsland(self, grid, r, c, rows, cols):
        if r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] == '0':
            return

        grid[r][c] = '0'
        
        directions = [[-1, 0], [1, 0], [0, 1], [0, -1]]

        for dr, dc in directions:
            self.sinkIsland(grid, r + dr, c + dc, rows, cols)

File: Data Structures & Algorithms/islands-and-treasure/submission-4.ts
class Solution {
    /**
     * @param {number[][]} grid
     */
    islandsAndTreasure(grid: number[][]): void {
        const ROWS = grid.length
        const COLS = grid[0].length

        const q: [number, number][] = [] //queue

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] == 0) {
                    q.push([r, c])
                } 
            }
        }

        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        const INF = 2147483647
        let distance = 0

        while (q.length) {
            const n = q.length
            for (let i = 0; i < n; i++) {
                const [r, c] = q.shift()

                for (const [dr, dc] of directions) {
                    const newRow = dr + r
                    const newCol = dc + c
                    if (newRow >= 0 && newCol >= 0 && newRow < ROWS && newCol < COLS && grid[newRow][newCol] === INF) {
                        grid[newRow][newCol] = distance + 1
                        q.push([newRow, newCol])
                    }
                }
            }

            distance++
        }
    }
}
